### PR TITLE
Add differentiable decomposition for max_pool2d/3d_with_indices

### LINKED ISSUE #85425: Deterministic Algorithm
### 🚀 The feature, motivation and pitch

I am working on an encoder that involves use of [torch.nn.MaxPool3d](https://pytorch.org/docs/stable/generated/torch.nn.MaxPool3d.html#torch.nn.MaxPool3d)
I was wondering if there can be a deterministic version for the same

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Kash6**
## Summary

Add a pure-tensor-op forward decomposition for `max_pool2d_with_indices` and `max_pool3d_with_indices` so that autograd computes the backward deterministically, no `atomicAdd` needed.

This picks up the approach from #175411 by @WizardsForgeGames, fixing the lint failures that were blocking it (PYREFLY errors from tuple defaults on `list[int]` parameters).

A generic `_max_pool_nd_with_indices` helper handles both 2D and 3D pooling by:
1. Padding the input with `-inf` so out-of-bounds positions never win the max
2. Building index tensors per spatial dimension for stride/dilation offsets
3. Using advanced indexing to gather all pooling windows via broadcasting
4. Taking the max over flattened kernel dimensions
5. Converting the local argmax back to flat indices into the original input

Inductor is unaffected as it has its own optimized max_pool decompositions that take priority. These global decompositions live outside `core_aten_decompositions()`, so they're only used by non-inductor backends and `torch.export`.

## Changes vs #175411
- Default params use `list[int] = []` / `[0]` / `[1]` with `# noqa: B006` instead of tuples, fixes PYREFLY lint
- Rebased on current main

## Testing
Verified on NVIDIA RTX 5080 (SM 12.0), PyTorch 2.11.0+cu128:
- Forward correctness: exact match (values and indices) against native impl across 11 configs (2D/3D, asymmetric kernels, dilation, ceil_mode, unbatched)
- Backward determinism: passes with `torch.use_deterministic_algorithms(True)` on CUDA
- Gradient consistency: identical across 5 runs
- Lint: clean (`lintrunner` passes)

Fixes #85425

cc @mruberry @kurtamohler @jansel @mikaylagawarecki

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -894,6 +894,11 @@ def max_pool3d_with_indices(
         )
     if stride is None:
         stride = torch.jit.annotate(list[int], [])
+    if not torch.jit.is_scripting():
+        if not input.is_cpu and torch.are_deterministic_algorithms_enabled():
```
Comment (kurtamohler) [inline torch/nn/functional.py]: Probably best to only do this for cuda

Comment (Kash6) [reply torch/nn/functional.py]: Forgive me for the oversight, good catch. @kurtamohler

**Comment by Kash6** [comment]
The CI runs seem to be a bit flaky, anything needed from my side? @kurtamohler @mikaylagawarecki

**Review by kurtamohler — APPROVED** [review]

**Comment by Kash6** [comment]
@kurtamohler  The `test_aten_core_operators` failures are from the expect file, I can't regenerate it locally without a source build. Could you run `EXPECTTEST_ACCEPT=1 python test/test_decomp.py HasDecompTest.test_aten_core_operators` and let me know what the expected content should be? Or if you can share the test output diff, I can update the file.

**Comment by kurtamohler** [comment]
@Kash6, please build and test your changes locally before submitting PRs. There's a contribution guide [here](https://github.com/pytorch/pytorch/wiki/The-Ultimate-Guide-to-PyTorch-Contributions#getting-started-with-pull-requests). Installation instructions are [here](https://github.com/pytorch/pytorch/blob/main/README.md#installation). If you don't have a machine with a CUDA compatible gpu, you can at least do a cpu-only build.

**Comment by Kash6** [comment]
@kurtamohler Apologies for the earlier iterations without local testing, my host OS (Ubuntu 25.10) has a glibc/CUDA incompatibility that prevented source builds. I've since set up a Docker-based build environment (Ubuntu 24.04 + CUDA 12.8) and have now verified all tests locally, including CUDA on my RTX 5080. All passing.

**Comment by Kash6** [comment]
There was a lint failure in the prior commit, fixed that. Lmk if anything else is needed @kurtamohler @jansel

**Comment by kurtamohler** [comment]
This will need an approval from someone on the appropriate reviewers list before being merging

**Review by jansel — APPROVED** [review]

**Comment by jansel** [comment]
@claude look into that CI failure
